### PR TITLE
Ignore excluded/skipped tests and unnecessary setup/teardown using SkipRemainingOnFailure

### LINKED ISSUE #2424: SkipRemainingOnFailure does not function correctly when there are nested BeforeAll blocks
### Checklist

- [X] Issue has a meaningful title
- [X] I have searched the existing issues. See [all issues](https://github.com/pester/Pester/issues?q=is%3Aissue)
- [X] I have tested using the latest version of Pester. See [Installation and update guide](https://pester.dev/docs/introduction/installation).

### What is the issue?

Pester v5.3.3
Invoke Pester with: `$config.Run.SkipRemainingOnFailure = "Container"`

If there are two describe blocks (A & B) that have nested BeforeAll steps
If the first Describe block(A) fails and hits and error.
The following desctibe block(B) should be completely skipped.
The issue is that the BeforeAll of the second Describe Block (B) is executed.

### Expected Behavior

If there are two describe blocks(A & B) that have nested BeforeAll steps
If the first Describe block(A) fails and hits and error.
The following describe block(B) should be completely skipped.
Specifically: The BeforeAll of the second Describe Block (B) should NOT be executed.

### Steps To Reproduce

_No response_

### Describe your environment

The following works - Example 1 without BeforeAll in Describe B:
Example 1 Code:
```powershell
Describe "A"{
    BeforeAll {
        Write-Host "Before ALL -A"
    }

    It "Test 1"{
        Write-Host "This is test 1"
        $a = 1
        $b = 2

        $a | Should -BeExactly $b
    }
}

Describe "B"{
     It "Test 2"{
        Write-Host "This is test 2"
    }
}
```
Example 1 - Output - without BeforeAll in Describe B:

```powershell
Running tests from 'C:\trial.Tests.ps1'
Before ALL -A
Describing A
This is test 1
  [-] Test 1 26ms (19ms|7ms)
   Expected exactly 2, but got 1.
   at $a | Should -BeExactly $b, C:\trial.Tests.ps1:11
   at <ScriptBlock>, C:\trial.Tests.ps1:11

Describing B
Skip: (B.Test 2) Test is skipped. 
  [!] Test 2 10ms (0ms|10ms)
Tests completed in 264ms
Tests Passed: 0, Failed: 1, Skipped: 1 NotRun: 0
Remaining tests skipped after first failure: 1
```

The following does not run as expected:
Example 2- code -  with BeforeAll in Describe B
```powershell
Describe "A"{
    BeforeAll {
        Write-Host "Before ALL -A"
    }

    It "Test 1"{
        Write-Host "This is test 1"
        $a = 1
        $b = 2

        $a | Should -BeExactly $b
    }
}

Describe "B"{
    BeforeAll {
        Write-Host "Before ALL -B"
    }

     It "Test 2"{
        Write-Host "This is test 2"
    }
}
```

Example 2: output -  with BeforeAll in Describe B
```powershell
Running tests from 'C:\trial.Tests.ps1'
Before ALL -A
Describing A
This is test 1
  [-] Test 1 30ms (21ms|8ms)
   Expected exactly 2, but got 1.
   at $a | Should -BeExactly $b, C:\trial.Tests.ps1:11
   at <ScriptBlock>, C:\trial.Tests.ps1:11
Before ALL -B

Describing B
Skip: (B.Test 2) Test is skipped. 
  [!] Test 2 11ms (0ms|11ms)
Tests completed in 288ms
Tests Passed: 0, Failed: 1, Skipped: 1 NotRun: 0
Remaining tests skipped after first failure: 1
```

### Possible Solution?

_No response_

### LINKED ISSUE #2440: SkipRemainingOnFailure reports skipped on filtered and explicitly skipped tests
### Checklist

- [X] Issue has a meaningful title
- [X] I have searched the existing issues. See [all issues](https://github.com/pester/Pester/issues?q=is%3Aissue)
- [X] I have tested using the latest version of Pester. See [Installation and update guide](https://pester.dev/docs/introduction/installation).

### What is the issue?

`Run.SkipRemainingOnFailure` processes tests that are both excluded by filter and/or skipped using `-Skip`. They are included in the summary

In the repro below, 3 tests were included by filter but one was skipped. The summary says `Remaining tests skipped after first failure: 3` when only 1 test was actually skipped by the plugin.

```console
Pester v5.6.0-beta1

Starting discovery in 1 files.
Discovery found 4 tests in 18ms.
Filter 'Tag' set to ('Demo').
Filters selected 3 tests to run.
Running tests.
Describing A
This is test 2
  [-] Test 1 11ms (4ms|7ms)
   Expected 2, but got 1.
   at 1 | Should -Be 2, untitled:Untitled-7:5
   at <ScriptBlock>, untitled:Untitled-7:5
 Describing B
   [!] Test 3 6ms (0ms|6ms)
   [!] Test 4 0ms (0ms|0ms)
Tests completed in 67ms
Tests Passed: 0, Failed: 1, Skipped: 2, Inconclusive: 0, NotRun: 1
Remaining tests skipped after first failure: 3
```

### Expected Behavior

Don't mark excluded tests as Skip.
Update plugin summary to only count skipped tests that it was responsible for (exclude `-Skip` tests).

### Steps To Reproduce

```powershell
$sb = {
    Describe 'A' {
        It 'Test 1' -Tag 'Demo' {
            Write-Host 'This is test 2'
            1 | Should -Be 2
        }
        It 'Test 2' {
            Write-Host 'This is test 3'
        }
        Describe 'B' {
            It 'Test 3' -Tag 'Demo' {
                Write-Host 'This is test 2'
            }
            It 'Test 4' -Tag 'Demo' -Skip {
                Write-Host 'This is test 3'
            }
        }
    }
}

$conf = New-PesterConfiguration
$conf.Output.Verbosity = 'Detailed'
$conf.Run.ScriptBlock = $sb
$conf.Run.SkipRemainingOnFailure = 'Block'
$conf.Filter.Tag = 'Demo'

Invoke-Pester -Configuration $conf
```

### Describe your environment

Pester version     : 5.6.0-beta1 /workspaces/Pester/bin/Pester.psm1                                                     
PowerShell version : 7.4.1
OS version         : Unix 5.15.133.1

### Possible Solution?

Update criteria from `if (-not $test.Executed)` to `if ($test.ShouldRun -and -not $test.Skip -and -not $test.Executed)`.

### PR DESCRIPTION
**Pull request by fflaten**
## PR Summary

- Refactors plugin to reduce code duplication.
- Ignores tests excluded by filters or explicitly skipped using `-Skip` from being marked as skipped by `Run.SkipRemainingOnFailure` option.
- Marks relevant blocks as skipped to avoid running `BeforeAll/AfterAll`

Fix #2440
Fix #2424

## PR Checklist

- [x] PR has meaningful title
- [x] Summary describes changes
- [x] PR is ready to be merged
  - If not, use the arrow next to `Create Pull Request` to mark it as a draft. PR can be marked `Ready for review` when it's ready.
- [x] Tests are added/update *(if required)*
- [ ] Documentation is updated/added *(if required)*

### CODE REVIEW

File: src/functions/Get-SkipRemainingOnFailurePlugin.ps1 [hunk 1]
```diff
@@ -1,21 +1,46 @@
...
+            $Context.Configuration.SkipRemainingOnFailureCount += 1
+            $test.Skip = $true
+            $test.ErrorRecord.Add($errorRecord)
+        }
+    } -OnBlock {
+        param($block)
+        if ($block.ShouldRun -and -not $block.Skip -and -not $block.Executed) {
+            # Marking remaining blocks as Skip to avoid executing BeforeAll/AfterAll
+            $block.Skip = $true
+        }
```
Comment (fflaten) [inline src/functions/Get-SkipRemainingOnFailurePlugin.ps1:41]: Should the block also get a ErrorRecord to explain why it was skipped?

File: src/functions/Get-SkipRemainingOnFailurePlugin.ps1 [hunk 2]
```diff
@@ -29,41 +54,18 @@ function Get-SkipRemainingOnFailurePlugin {
     $p.Start = {
         param ($Context)
 
+        # TODO: Use $Context.GlobalPluginData.SkipRemainingOnFailure.SkippedCount when exposed in $Context
```
Comment (nohwnd) [inline src/functions/Get-SkipRemainingOnFailurePlugin.ps1:57]: Is this something you will address in this PR or there should be issue created for this?

File: src/functions/Get-SkipRemainingOnFailurePlugin.ps1 (hunk 1 above)
Comment (nohwnd) [reply src/functions/Get-SkipRemainingOnFailurePlugin.ps1:41]: Not sure, do we write them all out?

File: src/functions/Get-SkipRemainingOnFailurePlugin.ps1 (hunk 2 above)
Comment (fflaten) [reply src/functions/Get-SkipRemainingOnFailurePlugin.ps1:57]: New issue. Didn't want to touch plugin interface as it wasn't critical.

I'm happy to provide a follow-up PR when name and steps to expose it in (all?) is decided.

File: src/functions/Get-SkipRemainingOnFailurePlugin.ps1 (hunk 1 above)
Comment (nohwnd) [reply src/functions/Get-SkipRemainingOnFailurePlugin.ps1:41]: Then you probably don't care if there are no tests, and if there are tests you will see the skipped tests, so imho adding the error is not necessary?

**Pull request merged**